ginga/table/AstroTable.py

```diff
@@ -0,0 +1,145 @@
+#
+# AstroTable.py -- Abstraction of an generic astro table.
+#
+# This is open-source software licensed under a BSD license.
+# Please see the file LICENSE.txt for details.
+#
+import logging
+
+import numpy as np
+from astropy.table import Table
+from astropy.io import fits
+
+from ginga.misc import Callback
+
+
+class TableError(Exception):
+    pass
+
+class AstroTable(Callback.Callbacks):
+
+    def __init__(self, data_ap=None, metadata=None, logger=None, name=None):
+
+        Callback.Callbacks.__init__(self)
+
+        if logger is not None:
+            self.logger = logger
+        else:
+            self.logger = logging.Logger('AstroTable')
+        self._data = data_ap
+        self.metadata = {}
+        self.name = name
+        # make sure a table has these attributes
+        self.metadata.setdefault('name', None)
+
+        # For callbacks
+        for name in ('modified', ):
+            self.enable_callback(name)
+
+    @property
+    def rows(self):
+        tab_a = self._get_data()
+        return len(tab_a)
+
+    @property
+    def columns(self):
+        tab_a = self._get_data()
+        return len(tab_a.colnames)
+
+    def get_size(self):
+        return (self.columns, self.rows)
+
+    def get_data(self):
+        return self._data
+
+    def _get_data(self):
+        return self._data
+
+    def get(self, kwd, *args):
+        if kwd in self.metadata:
+            return self.metadata[kwd]
+        else:
+            # return a default if there is one
+            if len(args) > 0:
+                return args[0]
+            raise KeyError(kwd)
+
+    def __getitem__(self, kwd):
+        return self.metadata[kwd]
+
+    def update(self, kwds):
+        self.metadata.update(kwds)
+
+    def set(self, **kwds):
+        self.update(kwds)
+
+    def __setitem__(self, kwd, value):
+        self.metadata[kwd] = value
+
+    def clear_metadata(self):
+        self.metadata = {}
+
+    def get_header(self):
+        # for compatibility with images--what should this contain?
+        return {}
+
+    def set_data(self, data_ap, metadata=None):
+        """Use this method to SHARE (not copy) the incoming table.
+        """
+        self._data = data_ap
+
+        if metadata:
+            self.update_metadata(metadata)
+
+        self.make_callback('modified')
+
+    def load_hdu(self, hdu, fobj=None):
+        self.clear_metadata()
+
+        try:
+            tbl = Table.read(hdu, format='fits')
+            self._data = tbl
+
+        except Exception as e:
+            self.logger.error("Error reading table from hdu: {0}".format(
+                str(e)))
+
+    def load_file(self, filepath, numhdu=None,
```

Comment:
Not sure if you ever plan to support anything other than FITS table (e.g., DAOPHOT, ECSV, general ASCII, etc), but if you do, `load_file()`'s API should be flexible enough to support other formats in the future.